design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,146 @@
+and allows making fine-grained access control.
+
+As Magento should provide [multiple BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends)
+it makes sense to extract authentication and authorization flow to a separate `Auth Service` to avoid credentials
+validation and duplication of roles & privileges data. And the general schema might look like this:
+
+![Authentication and Authorization General Schema](auth-general-schema.png)
+
+1. An unauthorized user sends a request with credentials to retrieve an access token, BFF/Monolith proxies this request to
+`Auth Service` which performs authorization & authentication and issues access token based on a user role.
+2. The user sends the request to get a needed resource with the access token, BFF/Monolith proxies this request to the appropriate
+service and there are few multiple strategies for token validation:
+    - BFF/Monolith just verifies the token signature (each service sends a token for validation to `Auth Service`)
+    - BFF/Monolith sends the token to `Auth Service` for validation
+3. A service sends the token for validation to `Auth Service` and, if a request is allowed, the service processes the request.
+4. In case, if the service requires additional data from another service, it creates a new request and set received token
+as in original request. So, additional service can send the token for validation to `Auth Service` to check user permissions.
+
+The basic scenario might be as following:
+
+![Auth Basic Scenario](auth-basic-scenario.png)
+
+A token validation per each request on `Auth Service` allows keeping authorization process in the centralized place, only
+`Auth Service` knows about all users, roles and assigned privileges. Each service during deployment should expose their API
+to `Auth Service`, which can automatically assign API to existing roles based on default rules. The token validation includes
+not only signature verification but authorization.
+
+### Services Deployment
+
+`Auth Service` stores allowed operations for roles and, during deployment, each service should send a list of available
+operations to `Auth Service`.
+
+![Services Deployment](services-deployment.png)
+
+### Service-to-Service Communication
+
+As each service should be agnostic to a source of a request (in general, for a service it does not matter who sends
+a request) service-to-service communication also requires a token. The next diagram describes a possible flow:
+
+![Service to Service](service-to-service.png)
+
+Before making a call to the needed service, the first service makes a call to `Auth Service` to retrieve token for the future
+communication. The second service makes a call for token validation in the same way as for other types of requests, so
+for the second service, it does not matter who sends original request.
+
+### Guest-to-Service Communication
```

Comment:
We also going to have service to monolith communication at the beginning. How would we distinguish these requests and enable our authorization mechanism for them?

Reply:
What kind of operation do you mean? When service should be an initiator for such request? In general, it's the same type of communication as service-to-service.

Reply:
For instance, Totals service after first iteration can still be part of monolith application. To add product to cart, Checkout service will need to send request to monolith application.

After more reading I noticed that you described this in the proposal, thank you.